File: 백준/Silver/9012. 괄호/괄호.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.Stack;

public class Main {
	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
		int T = Integer.parseInt(br.readLine());
		
		for (int t = 1; t <= T; t++) {
			String indent = br.readLine();
			boolean ans = true;
			Stack<Integer> stack = new Stack<>();
			
			for (int i = 0; i < indent.length(); i++) {
				if (indent.charAt(i) == '(')
					stack.push(1);
				else {
					if (stack.isEmpty()) {
						ans = false;
						break;
					}
					else 
						stack.pop();
				}
			}
			if (stack.isEmpty() && ans) bw.write("YES");
			else bw.write("NO");
			bw.newLine();
		}
		bw.flush();
		bw.close();
	}
}

File: 백준/Silver/9012. 괄호/괄호.py
T = int(input())
for _ in range(T):
    PR = input()
    ans = 0
    stop = 0
    for pr in PR:
        if pr == '(':
            ans += 1
        else:
            ans -= 1
        if ans < 0:
            stop = 1
            break
    if ans == 0 and stop == 0:
        print('YES')
    else:
        print('NO')